Steer/DigitizerWorkflow/src/CPVDigitizerSpec.cxx

```diff
@@ -0,0 +1,192 @@
+      return;
+    }
+    br->SetAddress(&mHitsBg);
+    br->GetEntry(entryID);
+  } else { //Bg
+    mHitsS->clear();
+    auto br = mSimChainS->GetBranch(brname);
+    if (!br) {
+      LOG(ERROR) << "No branch found";
+      return;
+    }
+    br->SetAddress(&mHitsS);
+    br->GetEntry(entryID);
+  }
+}
+
+void DigitizerSpec::run(framework::ProcessingContext& pc)
+{
+  if (mFinished)
+    return;
+
+  // read collision context from input
+  auto context = pc.inputs().get<o2::steer::RunContext*>("collisioncontext");
+  auto& timesview = context->getEventRecords();
+  LOG(DEBUG) << "GOT " << timesview.size() << " COLLISSION TIMES";
+
+  // if there is nothing to do ... return
+  if (timesview.size() == 0) {
+    return;
+  }
+
+  TStopwatch timer;
+  timer.Start();
+
+  LOG(INFO) << " CALLING CPV DIGITIZATION ";
+  std::vector<TriggerRecord> triggers;
+  static std::vector<o2::cpv::Hit> hits;
+
+  mLabels.clear();
+  mDigits.clear();
+  int indexStart = mDigits.size();
+  auto& eventParts = context->getEventParts();
+  // loop over all composite collisions given from context
+  // (aka loop over all the interaction records)
+  for (int collID = 0; collID < timesview.size(); ++collID) {
+    mDigitizer.setEventTime(timesview[collID].timeNS);
+
+    // for each collision, loop over the constituents event and source IDs
+    // (background signal merging is basically taking place here)
+    for (auto& part : eventParts[collID]) {
+
+      // get the hits for this event and this source
+      retrieveHits("CPVHit", part.sourceID, part.entryID);
+      mDigitizer.setCurrEvID(part.entryID);
+    }
+
+    LOG(DEBUG) << "Found " << mHitsBg->size() << " BG hits and " << mHitsS->size() << "signal hits";
+
+    // call actual digitization procedure
+    mDigitizer.process(mHitsBg, mHitsS, mDigits, mLabels);
+
+    // Add trigger record
+    triggers.emplace_back(timesview[collID], indexStart, mDigits.size() - indexStart);
+    indexStart = mDigits.size();
+
+    LOG(DEBUG) << "Have " << mDigits.size() << " digits ";
+  }
+
+  LOG(DEBUG) << "Have " << mLabels.getNElements() << " CPV labels ";
+  // here we have all digits and we can send them to consumer (aka snapshot it onto output)
+  pc.outputs().snapshot(Output{"CPV", "DIGITS", 0, Lifetime::Timeframe}, mDigits);
+  pc.outputs().snapshot(Output{"CPV", "DIGITTRIGREC", 0, Lifetime::Timeframe}, triggers);
+  pc.outputs().snapshot(Output{"CPV", "DIGITSMCTR", 0, Lifetime::Timeframe}, mLabels);
+  // CPV is always a triggering detector
+  const o2::parameters::GRPObject::ROMode roMode = o2::parameters::GRPObject::TRIGGERING;
+  LOG(DEBUG) << "CPV: Sending ROMode= " << roMode << " to GRPUpdater";
+  pc.outputs().snapshot(Output{"CPV", "ROMode", 0, Lifetime::Timeframe}, roMode);
+
+  timer.Stop();
+  LOG(INFO) << "Digitization took " << timer.CpuTime() << "s";
+
+  // we should be only called once; tell DPL that this process is ready to exit
+  pc.services().get<ControlService>().readyToQuit(QuitRequest::Me);
```

Comment:
please add ``pc.services().get<of::ControlService>().endOfStream();`` to not block potentially dependent devices